### PR TITLE
Cleaner Solution Code for Xenia and Tree

### PR DESCRIPTION
**Pull request by dutinmeow**
The previous solution was pretty blocky, so I sorted everything out into functions and added comments.

### CODE REVIEW

File: content/5_Plat/Centroid.mdx [hunk 1]
```diff
@@ -235,99 +235,142 @@ private static class Centroid {
...
-std::vector<info> g[MN];
-
-int dfs(int n, int p=0)
-{
-	s[n]=1;
-	for(auto x:a[n])
-		if(x!=p&&!r[x])
-			s[n]+=dfs(x, n);
-	return s[n];
+#pragma GCC optimize ("Ofast")
```
Comment (bqi343) [inline content/5_Plat/Centroid.mdx:244]: > #pragma GCC optimize ("Ofast")

don't include these, they're not relevant to the solution

File: content/5_Plat/Centroid.mdx [hunk 2]
```diff
@@ -235,99 +235,142 @@ private static class Centroid {
...
-			t=n, b=v[n].size(), d=1;
-			dfs2(x, n);
-			v[n].push_back(INF);
-		}
-	r[n]=1;
-	for(auto x:a[n])
-		if(!r[x])
-			centroid(x);
+ 
+// this stuff is for tree distance, see binary jumping mondule
```
Comment (bqi343) [inline content/5_Plat/Centroid.mdx:292]: there's no need to use binary jumping

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Centroid.mdx [hunk 3]
```diff
@@ -235,99 +235,142 @@ private static class Centroid {
...
-{
-	for(auto x:a[n])
-		if(!r[x]&&x!=p&&s[x]*2>ms)
-			return find(x, ms, n);
-	return n;
+ 
+int centroid(int u, int n, int p = 0) {
+    for (int v : adj[u])
+        if (v != p && !r[v])
+            if (W[v] * 2 > n)
```
Comment (bqi343) [inline content/5_Plat/Centroid.mdx:278]: also tabs

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Centroid.mdx (hunk 2 above)
Comment (dutinmeow) [reply content/5_Plat/Centroid.mdx:289]: How would you do it without binary jumping?

Comment (bqi343) [reply content/5_Plat/Centroid.mdx:289]: solution on the left doesn't use binary jumping

Comment (dutinmeow) [reply content/5_Plat/Centroid.mdx:289]: Both tutorials linked on the article use the binary jumping/lca method, so I think its more intuitive

File: content/5_Plat/Centroid.mdx [hunk 4]
```diff
@@ -235,99 +235,139 @@ private static class Centroid {
...
+// ans[] is closest red node in subtree of centroid tree
+// adj[] stores original tree given in input
+// col[] stores if node has already been colored red (optional)
+ 
+bool r[100005];
+// r[] stores if a node has been deleted
+ 
+int dfs_w(int u, int p = 0) {
+	if (r[u])
+		return 0;
```
Comment (frodakcin) [inline content/5_Plat/Centroid.mdx:264]:
```suggestion
	if (r[u])
		return 0;
```

This is redundant, right? Probably can delete.

**Review by frodakcin — COMMENTED** [review]
Looks fine. Would it be better to cut out extra log factor though?

File: content/5_Plat/Centroid.mdx (hunk 4 above)
Comment (dutinmeow) [reply content/5_Plat/Centroid.mdx]: > Looks fine. Would it be better to cut out extra log factor though?

Both Galen Colin's tutorial and the other one mention using some method to find distance. I could use sparse table to do it in constant time, but I feel like binary jumping is just more intuitive from an educational standpoint, seeing as it's part of the platinum topics. In the end, it really doesn't matter as long as the solution pass the tests.

Comment (thecodingwizard) [reply content/5_Plat/Centroid.mdx]: maybe it suffices to just mention that it's possible to cut out the extra log factor? @frodakcin @bqi343

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Centroid.mdx (hunk 4 above)
Comment (bqi343) [reply content/5_Plat/Centroid.mdx]: Just do it in the same way as the original solution does? It's quite simple and there's no extra log factor.

File: content/5_Plat/Centroid.mdx (hunk 2 above)
Comment (bqi343) [reply content/5_Plat/Centroid.mdx:287]: oops started this comment a few days ago but forgot to submit it

Comment (dutinmeow) [reply content/5_Plat/Centroid.mdx:287]: I'll work on the implementation, but I'm quite busy the next few days due to schoolwork.

**Comment by bqi343** [comment]
closing as desired changes have not been made

**Pull request closed without merging**